### ISSUE TITLE
Self-Service Migrations

### ISSUE THREAD

**Issue by dylan-smith** [github/gh-gei#101]
Bring the Octoshift CLI to a state that aligns with how customers want to use Octoshift, covers running migrations from ADO, GHES & GHEC, and can be picked up self-service and used without in hand holding for the first few migrations. This will help to reduce the time investment that's needed to get a customer started using Octoshift. While also bringing the CLI benefit to GHES and GHEC migration customers.

DRI: @dylan-smith 

## Batches

- [x] #103 
- [x] #104 
- [x] #105 
- [x] #106

**Comment by dylan-smith** [comment]
Status: 🟢

Target date: 2022-01-31

Work done this week
- Created a new "gei" CLI for GitHub -> GitHub migrations. Tested with GHEC -> GHEC
  - #54
  - #55 
  - #192 (expected to land tomorrow/monday)
- Add --ssh flag to allow customers who are encountering problems related to the HTTPS change to fall back to the old behavior
  - #193 
- Masking secrets from the logs to prevent accidental PAT disclosure when sharing log files
  - #123 
- Continued improvements in our unit test coverage, up to 81% coverage
  - #114 
![image](https://user-images.githubusercontent.com/1508559/149418085-634b2da8-ed91-42fe-9eb0-37c182dfee9b.png)


Work planned next week
- v0.6 releasing on Monday with the above features
- Show and Tell with the MVP's on Thursday
- Support for paging when calling GitHub API's. The main place this will be noticed is when running generate-script against a GitHub source. The CLI currently will only list the first 30 repos because we haven't implemented paging like we have for ADO.
  - #45 
- Improved error handling. User friendly messages + proper error codes
  - #116 
- Begin work on e2e tests

**Comment by dylan-smith** [comment]
### Status: 🟢

### Target date: 2022-01-31

### Work done this week
- v0.6 released on Monday
  - Release Notes: https://github.com/github/gh-gei/releases/tag/v0.6
- Show and Tell with the MVP's
  - [Recording](https://microsoft-my.sharepoint.com/personal/v-suwild_microsoft_com/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fv%2Dsuwild%5Fmicrosoft%5Fcom%2FDocuments%2FRecordings%2FAzure%20DevOps%20%2D%5F%20GitHub%20Migration%20Tooling%2D20220120%5F090209%2DMeeting%20Recording%2Emp4&parent=%2Fpersonal%2Fv%2Dsuwild%5Fmicrosoft%5Fcom%2FDocuments%2FRecordings) 
- Improved error handling. User friendly messages + proper error codes
  - #116 
- Publish gei as an extension to the official GitHub CLI
  - #127 
- Begin work on e2e tests
  - #3 
- Continuing to improve Unit Test coverage. Up to 86% from 81% last week
  - Only one major piece of code remaining before we can call this done
  - #114 
![image](https://user-images.githubusercontent.com/1508559/150467869-a1626801-2726-44c4-9b2b-6f367938aa53.png)


## Next week
- Support for paging when calling GitHub API's. The main place this will be noticed is when running generate-script against a GitHub source. The CLI currently will only list the first 30 repos because we haven't implemented paging like we have for ADO.
  - #45 
- Finish the e2e testing efforts
  - #3 
- Support for ADO source in the gh->gh CLI
  - #125 
  - #126 
- migrator role commands in the gh->gh CLI
  - #207 
  - #208 
- GHES as a source for gh->gh CLI
  - #128 
- Parallel migration support
  - #106 
- CLI Telemetry
  - #107

**Comment by dylan-smith** [comment]
### Status: Green
Updated target date estimate to Feb 18th.

### Target date: 2022-02-18

### Work done this week
- Finish the e2e testing efforts
  - #3 
- Support for ADO source in the gh->gh CLI
  - #125 
  - #126 
- migrator role commands in the gh->gh CLI
  - #207 
  - #208 

### Work planned next week
- GHES as a source for gh->gh CLI
  - #128 
- Parallel migration support
  - #106

**Comment by dylan-smith** [comment]
### Status: Green

### Target date: 2022-03-04

### Work done this week
- GHES as a source for gh->gh CLI
  - #128 
- Parallel migration support was started, but Arin is OOO this work so was put on hold until he returns
  - #106 

### Work planned next week
- Parallel migration support implementation continues
  - #106 
- Telemetry
  - #107

**Comment by dylan-smith** [comment]
### Status: Green

### Target date: 2022-03-04

### Work done this week
- Parallel migration support is progressing, still probably another week of work remains
  - #106 

### Work planned next week
- Parallel migration support implementation continues, possibly completes next week
  - #106 
- Telemetry
  - #107

**Comment by dylan-smith** [comment]
### Status: Green

### Target date: 2022-03-31

### Work done this week
- Parallel migration work is code complete, waiting on PR reviews before we can release it
  - #106 

### Work planned next week
- Telemetry
  - #107

**Comment by dylan-smith** [comment]
### Status: Green

### Target date: 2022-03-31

### Work done this week
- Parallel migration work was merged in, but a number of issues were discovered during final release testing that we're still working on resolving
  - #106 
  - #254 
  - #255 
  - #264 
  - #269 
- Work started on making the CLI commands idempotent
  - #259 

### Work planned next week
- Telemetry
  - #107 
- Continue work on idempotency
  - #117

**Comment by dylan-smith** [comment]
### Status: Green (Complete)

Moved the telemetry batch out of this epic. This is now complete.

### Target date: 2022-03-31

### Work done this week
- Parallel migration support shipped in v0.10
  - #106 
- Idempotency making more progress
  - #287 
- API updates to support backend changes
  - #296 
  - #258
- --ado-team-project option added for scoping (external contribution)
  - #124 
  - #242 
- Last task for missing unit tests complete, our code coverage looks great now!
  - #177 
![image](https://user-images.githubusercontent.com/1508559/161258896-e86eca26-ce9c-4dbd-90f6-d41e6e568356.png)
- Added support for M1 macs
  - #302 
- Re-enabled integration tests after the underlying issues causing flakiness have been resolved
  - #300
